File: frontend/js/modules/permissions.js
// Role -> permission mapping. Mirrors backend/services/permissions.py — keep
// both in sync. Used to gate UI elements on the active user's role.
import { store } from '../utils/state.js';

const PERMISSIONS = {
  admin: new Set([
    'upload_dataset', 'delete_dataset', 'rename_dataset', 'edit_dataset',
    'create_dashboard', 'create_chart', 'duplicate_any_chart',
    'save_query_template', 'export_chart', 'apply_filters',
    'manage_users', 'view_audit_log', 'change_user_role',
  ]),
  editor: new Set([
    'edit_dataset', 'create_dashboard', 'create_chart',
    'duplicate_any_chart', 'save_query_template',
    'export_chart', 'apply_filters',
  ]),
  viewer: new Set([
    'export_chart', 'apply_temp_filters',
  ]),
};

export function can(permission) {
  const u = store.get('activeUser');
  return !!u && PERMISSIONS[u.role]?.has(permission);
}

function isOwner(resource) {
  const u = store.get('activeUser');
  return !!u && resource && resource.created_by === u.id;
}

export function canEditChart(chart) {
  const u = store.get('activeUser');
  if (!u || !chart) return false;
  return u.role === 'admin' || (u.role === 'editor' && isOwner(chart));
}

// Removes elements that the current user cannot access, based on a
// `data-requires-permission="<perm>"` attribute. Call after every render.
export function applyPermissions(rootEl) {
  if (!rootEl) return;
  rootEl.querySelectorAll('[data-requires-permission]').forEach(el => {
    if (!can(el.dataset.requiresPermission)) el.remove();
  });
}

File: backend/services/permissions.py
"""Role -> permission mapping. Mirrored on the frontend at
frontend/js/modules/permissions.js — keep both in sync.
"""

PERMISSIONS = {
    "admin": {
        "upload_dataset", "delete_dataset", "rename_dataset", "edit_dataset",
        "create_dashboard", "create_chart", "duplicate_any_chart",
        "save_query_template", "export_chart", "apply_filters",
        "manage_users", "view_audit_log", "change_user_role",
    },
    "editor": {
        "edit_dataset", "create_dashboard", "create_chart",
        "duplicate_any_chart", "save_query_template",
        "export_chart", "apply_filters",
    },
    "viewer": {
        "export_chart", "apply_temp_filters",
    },
}


def can(user_role: str, permission: str) -> bool:
    return permission in PERMISSIONS.get(user_role, set())


def _is_owner(user: dict, resource: dict) -> bool:
    return bool(user) and resource.get("created_by") == user.get("id")


def can_edit_chart(user: dict, chart: dict) -> bool:
    if not user:
        return False
    return user["role"] == "admin" or (user["role"] == "editor" and _is_owner(user, chart))


def can_delete_chart(user: dict, chart: dict) -> bool:
    return can_edit_chart(user, chart)


def can_edit_dashboard(user: dict, dashboard: dict) -> bool:
    if not user:
        return False
    return user["role"] == "admin" or (user["role"] == "editor" and _is_owner(user, dashboard))


def can_delete_dashboard(user: dict, dashboard: dict) -> bool:
    return can_edit_dashboard(user, dashboard)
